crates/api/src/kitsune.rs

```diff
@@ -0,0 +1,70 @@
+//! Kitsune2 top-level module related types.
+
+use crate::*;
+use std::sync::Arc;
+
+/// Handler for events coming out of Kitsune2.
+pub trait KitsuneHandler: 'static + Send + Sync + std::fmt::Debug {
+    /// Gather preflight data to send to a new opening connection.
+    /// Returning an Err result will close this connection.
+    ///
+    /// The default implementation sends an empty preflight message.
+    fn preflight_gather_outgoing(
+        &self,
+        peer_url: String,
+    ) -> K2Result<bytes::Bytes> {
+        drop(peer_url);
+        Ok(bytes::Bytes::new())
+    }
+
+    /// Validate preflight data sent by a remote peer on a new connection.
+    /// Returning an Err result will close this connection.
+    ///
+    /// The default implementation ignores the preflight data,
+    /// and considers it valid.
+    fn preflight_validate_incoming(
+        &self,
+        peer_url: String,
+        data: bytes::Bytes,
+    ) -> K2Result<()> {
+        drop(peer_url);
+        drop(data);
+        Ok(())
+    }
+
+    /// Kitsune would like to construct a space. Provide a handler.
+    fn create_space(
```

Comment:
This is a hook so that the user can customize what pieces get used by a space? Wouldn't that be decided by what factories are plugged into already? Or are we thinking that some spaces might be set up differently to others?

This just seems slightly strange to me so I'm wanting to understand where the responsibilities are for setting up a space

Reply:
The space is the level at which messaging between peers is received. So the handler that does the accepting of those messages is the SpaceHandler. This hook at the top-level handler is so that conductor or whatever other implementor can then construct those SpaceHandlers to be able to receive messages from remote peers.

Reply:
Perhaps it could help if you added some integration test stub that shows how this would work with Holochain being the consumer.

Reply:
Done: https://github.com/holochain/kitsune2/pull/40/commits/fa0715ba92c9dd001b8e27629820ad5e392e3f5b

Reply:
This is like the equivalent of the "add a new messaging handle" with the ghost actor in the current implementation, if I've understood this right :)

Reply:
Yeah, exactly, when conductor hooks up the "event" type channel so it can receive events out of legacy kitsune.

Reply:
@neonphog Thanks for the test!